packages/hdwallet-provider/src/index.ts

```diff
@@ -28,6 +28,228 @@ import { Callback, JsonRPCResponse } from "web3/providers";
 // See issue #65 for more
 const singletonNonceSubProvider = new NonceSubProvider();
 
+/*
+ * type aliases for better readability around legacy positional arguments
+ */
+export type MnemonicPhrase = string;
+export type MnemonicPassword = string;
+export interface Mnemonic {
+  phrase: MnemonicPhrase;
+  password?: MnemonicPassword;
+}
+export type PrivateKey = string;
+export type Provider = any;
+export type ProviderUrl = string;
+export type ProviderOrUrl = Provider | ProviderUrl;
+export type AddressIndex = number;
+export type NumberOfAddresses = number;
+export type ShareNonce = boolean;
+export type DerivationPath = string;
+
+/*
+ * namespace wrapper for new constructor options interface
+ */
+export namespace NewConstructor {
+  export interface MnemonicSigningAuthority {
+    mnemonic: Mnemonic;
+  }
+
+  export interface PrivateKeySigningAuthority {
+    privateKeys: PrivateKey[];
+  }
+
+  export type SigningAuthority =
+    | MnemonicSigningAuthority
+    | PrivateKeySigningAuthority;
+
+  export interface CommonOptions {
+    providerOrUrl: ProviderOrUrl;
+    addressIndex?: AddressIndex;
+    numberOfAddresses?: NumberOfAddresses;
+    shareNonce?: ShareNonce;
+    derivationPath?: DerivationPath;
+  }
+
+  export type Options = SigningAuthority & CommonOptions;
+
+  // extract the mnemonic if that's the style used, or return undefined
+  export const getMnemonic = (
+    signingAuthority: SigningAuthority
+  ): Mnemonic | undefined => {
+    if ("mnemonic" in signingAuthority) {
+      return signingAuthority.mnemonic;
+    }
+  };
+
+  // extract the private keys if that's the style used, or return undefined
+  export const getPrivateKeys = (
+    signingAuthority: SigningAuthority
+  ): PrivateKey[] | undefined => {
+    if ("privateKeys" in signingAuthority) {
+      return signingAuthority.privateKeys;
+    }
+  };
+}
+
+/*
+ * namespace wrapper for old-style positional arguments
+ */
+export namespace LegacyConstructor {
+  type PossibleArguments = [
+    /*
+     * required
+     */
+    MnemonicPhrase | PrivateKey[],
```

Comment:
> so I think there's no harm

caveat: we'll need to make sure that this branch of code is hit, even with the new-style constructor, since I don't want to accidentally lose the error message for this situation.

Reply:
@gnidan, feel free to "steal the fun" if you get a chance. I probably won't have time to look at this PR again until next week at the earliest.

Reply:
I've also got something of a busy week, so we'll see! Guess it's a race, then!